android/capacitor/src/main/java/com/getcapacitor/CapConfig.java

```diff
@@ -52,81 +116,341 @@ private void loadConfig(AssetManager assetManager) {
+     */
+    private void deserializeConfig() {
+        // Server
+        html5mode = getBoolean(configJSON, "server.html5mode", html5mode);
+        serverUrl = getString(configJSON, "server.url", null);
+        hostname = getString(configJSON, "server.hostname", hostname);
+        androidScheme = getString(configJSON, "server.androidScheme", androidScheme);
+        allowNavigation = getArray(configJSON, "server.allowNavigation", null);
+
+        // Android
+        overriddenUserAgentString = getString(configJSON, "android.overrideUserAgent", getString(configJSON, "overrideUserAgent", null));
+        appendedUserAgentString = getString(configJSON, "android.appendUserAgent", getString(configJSON, "appendUserAgent", null));
+        backgroundColor = getString(configJSON, "android.backgroundColor", getString(configJSON, "backgroundColor", null));
+        allowMixedContent =
+            getBoolean(configJSON, "android.allowMixedContent", getBoolean(configJSON, "allowMixedContent", allowMixedContent));
+        captureInput = getBoolean(configJSON, "android.captureInput", captureInput);
+        webContentsDebuggingEnabled = getBoolean(configJSON, "android.webContentsDebuggingEnabled", webContentsDebuggingEnabled);
+        hideLogs = getBoolean(configJSON, "android.hideLogs", getBoolean(configJSON, "hideLogs", hideLogs));
+
+        // Plugins
+        pluginConfigurations = getObject(configJSON, "plugins");
+    }
+
+    public boolean html5mode() {
+        return html5mode;
+    }
+
+    public String getServerUrl() {
+        return serverUrl;
+    }
+
+    public String getHostname() {
+        return hostname;
+    }
+
+    public String getAndroidScheme() {
+        return androidScheme;
+    }
+
+    public String[] getAllowNavigation() {
+        return allowNavigation;
+    }
+
+    public String getOverriddenUserAgentString() {
+        return overriddenUserAgentString;
+    }
+
+    public String getAppendedUserAgentString() {
+        return appendedUserAgentString;
+    }
+
+    public String getBackgroundColor() {
+        return backgroundColor;
+    }
+
+    public boolean allowMixedContent() {
+        return allowMixedContent;
+    }
+
+    public boolean captureInput() {
+        return captureInput;
+    }
+
+    public Boolean getWebContentsDebuggingEnabled() {
+        return webContentsDebuggingEnabled;
+    }
+
+    public boolean hideLogs() {
+        return hideLogs;
+    }
+
+    public JSONObject getPluginConfigurations() {
+        return pluginConfigurations;
+    }
+
+    /**
+     * Get a string value for a plugin in the Capacitor config.
+     *
+     * @param pluginId The ID of the plugin used in the config
+     * @param configKey The key of the value to retrieve
+     * @return The value from the config, if exists. Null if not
+     */
+    public String getPluginString(String pluginId, String configKey) {
+        return getPluginString(pluginId, configKey, null);
+    }
```

Comment:
I don't think we should add these methods here. Is the intended usage `this.bridge.config.getPluginString()` in the Plugin class? If so, it would be a nicer API if it was just `this.config.getString()` (or `this.pluginConfig.getString()`), etc. We could make a `PluginConfig` class to keep classes smaller and purposeful.

Reply:
> CapConfig was created to avoid a breaking change as we had the Config class and users were using it on plugins. Used CapConfig as the name to also match iOS name.
> Since this PR changes the constructor, that's also a breaking change.
> Not sure if instead of changing the constructor, we rename the class to `InstanceConfiguration` as Ian did on the iOS equivalent #3759

It is a breaking change, but is it expected that 3rd parties would be creating new instances of CapConfig? I updated the code in the core to use the new constructor. If it's not an expected use-case or an embedded/edge case, it might be an acceptable breaking change.

I can see the appeal in naming things similarly to iOS, but I don't think it should be the rule. Java has instances of classes so the name "InstanceConfiguration" feels unnecesarily verbose, but we are likely introducing a new class here to contain the plugin config, so we do need some name work.

Reply:
I split the Plugin JSON config into its own class and moved the methods there also. Since the CapConfig and PluginConfig class used similar JSON parsing methods, I centralized those in a new class called JSONUtils that both Config classes may use. Thoughts?

Reply:
Looking good so far--check out my latest comment: https://github.com/ionic-team/capacitor/pull/3778#discussion_r540597530